api/app/tests/fba/test_fba_error_random_sample.py

```diff
@@ -0,0 +1,239 @@
+
+
+acceptable_margin_of_error: Final = 0.01
+
+
+def check_metric(metric: str,
+                 fuel_type: str,
+                 python_value: float,
+                 comparison_value: float,
+                 metric_error_margin: float, note: str = None):
+    """ Check relative error of a metric """
+    # logging with %s became unreadable:
+    # pylint: disable=logging-fstring-interpolation
+    if comparison_value < 0:
+        logger.warning('Skipping %s! (%s) - note: %s', metric, comparison_value, note)
+    else:
+        assert python_value >= 0
+        error = relative_error(python_value, comparison_value)
+        absolute_error = abs(python_value - comparison_value)
+        # logger.info(
+        #     f'{fuel_type}: Python {python_value}, {metric} {comparison_value} ; error: {error}')
+        if error > acceptable_margin_of_error:
+            logger.error('%s %s relative error %s > %s! (python: %s, expected: %s)',
+                         fuel_type, metric, error, acceptable_margin_of_error, python_value, comparison_value)
+        if metric_error_margin > 0.01:
+            logger.debug('%s: The acceptable margin of error (%s) for %s is set too high',
+                         fuel_type, metric_error_margin, metric)
+        if error > metric_error_margin:
+            # ok - fine, but for small numbers this isn't a big deal, so let's check the absolure difference.
+            if absolute_error < 0.01:
+                logger.debug('no big deal, the absolute difference (%s) is tiny!', absolute_error)
+            else:
+                assert error < metric_error_margin, f'{fuel_type}:{metric}'
+
+
+@pytest.mark.usefixtures('mock_jwt_decode')
+@scenario('test_fba_error_random_sample.feature', 'Fire Behaviour Calculation',
+          example_converters=dict(crown_base_height=_str2float,
+                                  fuel_type=str,
+                                  percentage_conifer=_str2float,
+                                  percentage_dead_balsam_fir=_str2float,
+                                  grass_cure=_str2float,
+                                  num_iterations=int,
+                                  ros_margin_of_error=float,
+                                  hfi_margin_of_error=float,
+                                  cfb_margin_of_error=float,
+                                  one_hour_spread_margin_of_error=float))
+def test_fire_behaviour_calculator_scenario():
+    """ BDD Scenario. """
+
+
+@given("""<fuel_type>, <percentage_conifer>, <percentage_dead_balsam_fir>, <grass_cure> and """
+       """<crown_base_height> for <num_iterations>""",
+       target_fixture='results')
+def given_input(fuel_type: str, percentage_conifer: float, percentage_dead_balsam_fir: float,
+                grass_cure: float, crown_base_height: float, num_iterations: int):
+    """ Take input and calculate actual and expected results """
+
+    # get python result:
+    random.seed(42)
```

Comment:
42 is NOT the answer to everything!

Reply:
:)

Reply:
I think it would be better not supplying a seed, so that it seeds based on current system time: https://docs.python.org/3/library/random.html#random.seed

Otherwise we'll have the same series of random numbers every time this is run. With a different seed every time these tests are run we're more likely to run into a test case failure.

Reply:
Not using a seed makes it practically un-reproducible though, and will result in intermittent failure. I would much rather have a deterministic set of unit tests. I certainly see the merits, but also see someone just re-running the test to get it to pass with a different set of random conditions.

Reply:
We should get the specific value that we can reproduce the failure with when the test fails.

Reply:
But if we don't want to investigate any new failures it makes sense.

Reply:
We could totally log the seed, so that someone can go back, plug the same seed in, and re-produce the error. Let's chat today - it's a good idea. I'm just worried that no-one is going to follow up and rather just pres "re-run".

Reply:
Either way - I've got another PR coming in right after this one - we'll chat and then I can put a changing seed in there.

Reply:
Ok sounds good!